traced/ykrt/src/mt.rs

```diff
@@ -139,6 +142,37 @@ impl MT {
             }
         })
     }
+
+    /// Queue `job` to be run on a worker thread.
+    fn queue_job(&self, job: Box<dyn FnOnce() + Send>) {
+        // We have a very simple model of worker threads. Each time a job is queued, we spin up a
+        // new worker thread iff we aren't already running the maximum number of worker threads.
+        // Once started, a worker thread never dies, and spins endlessly waiting for work.
+
+        let inner = Arc::clone(&self.inner);
+        inner.job_queue.lock().unwrap().push_back(job);
+
+        let max_jobs = inner.max_worker_threads.load(Ordering::Relaxed);
+        if inner.active_worker_threads.load(Ordering::Relaxed) < max_jobs {
+            // At the point of the `load` on the previous line, we weren't running the maximum
+            // number of worker threads. There is now a possible race condition where multiple
+            // threads calling `queue_job` could try creating multiple worker threads and push us
+            // over the maximum worker thread limit.
+            if inner.active_worker_threads.fetch_add(1, Ordering::Relaxed) > max_jobs {
+                // Another thread(s) is also spinning up another worker thread and they won the
+                // race.
+                inner.active_worker_threads.fetch_sub(1, Ordering::Relaxed);
+                return;
+            }
+
+            let inner_cl = Arc::clone(&self.inner);
+            thread::spawn(move || loop {
```

Comment:
Do we only check if we can assign a job to a worker thread at the time that a job is queued then?

Reply:
Wait... It spins forever even if there is no work?

So if we have at some time in the past had 8 threads compiling simultaneously, then we will forever have 8 threads spinning looking for work?

Reply:
> Do we only check if we can assign a job to a worker thread at the time that a job is queued then?

No, we spin worker threads up each time a new job is added, but it's possible that by the time they've started working that another worker thread has picked the job up.

> Wait... It spins forever even if there is no work?

No, they gracefully wait (https://github.com/softdevteam/yk/pull/314/commits/b3c483dcb0817efd121ed4f613db881c3d3f8090).